### ISSUE TITLE
PR: feature: Implement caddyfile.Unmarshaler interface

### ISSUE THREAD

**Issue by golgeek** [greenpau/caddy-auth-portal#149]
I'm currently writing a Caddy V2 module that replicates the behavior of [Fabio](https://github.com/fabiolb/fabio).

In a nutshell, my module connects to Consul, fetches the services to reverse-proxy, generates Caddy's `http` app configuration for and keeps it up-to-date by watching the Consul services and regenerating it when necessary (new services, health issues on backends, ...).

As I want to have authentication on certain subdomains, I need to encapsulate the `caddy-auth-portal` configuration in my own module's configuration to be able to generate the final Caddy configuration (as I generate the `http` app configuration from scratch).

This currently works well in JSON (as Caddy automatically unmarshals the JSON configuration in the the declared struct), but the implementation of the caddyfile.Unmarshaler interface is required to be able to have the same behavior with the Caddyfile format.

With the PR #148, in my own Caddyfile parsing, I can now call `caddyfile.UnmarshalModule(d, "http.handlers.authp")`, which more or less gives me an instance of `authn.Authenticator{}` that I can use in my own codebase.

---

The patch is very lightweight.
It splits the current `parseCaddyfileAuthenticator()` into two subs:
- one that creates the `authn.Authenticator` instance from the Caddyfile format
- one (`getRouteFromParseCaddyfileAuthenticator()`) that wraps it in the new `caddyhttp.Route`

Then, the new `UnmarshalCaddyfile(d *caddyfile.Dispenser)` func only calls `parseCaddyfileAuthenticator`.

Patch has been tested, and unit tests are still running:
```
ok  	github.com/greenpau/caddy-auth-portal	11.742s
ok  	github.com/greenpau/caddy-auth-portal/internal/tag	(cached)
?   	github.com/greenpau/caddy-auth-portal/internal/tests	[no test files]
?   	github.com/greenpau/caddy-auth-portal/internal/utils	[no test files]
ok  	github.com/greenpau/caddy-auth-portal/pkg/authn	1.025s
ok  	github.com/greenpau/caddy-auth-portal/pkg/backends	(cached)
ok  	github.com/greenpau/caddy-auth-portal/pkg/backends/ldap	(cached)
ok  	github.com/greenpau/caddy-auth-portal/pkg/backends/local	(cached)
?   	github.com/greenpau/caddy-auth-portal/pkg/backends/oauth2	[no test files]
?   	github.com/greenpau/caddy-auth-portal/pkg/backends/saml	[no test files]
ok  	github.com/greenpau/caddy-auth-portal/pkg/cache	(cached)
ok  	github.com/greenpau/caddy-auth-portal/pkg/cookie	(cached)
?   	github.com/greenpau/caddy-auth-portal/pkg/enums/operator	[no test files]
?   	github.com/greenpau/caddy-auth-portal/pkg/errors	[no test files]
?   	github.com/greenpau/caddy-auth-portal/pkg/registration	[no test files]
ok  	github.com/greenpau/caddy-auth-portal/pkg/transformer	(cached)
ok  	github.com/greenpau/caddy-auth-portal/pkg/ui	0.066s
ok  	github.com/greenpau/caddy-auth-portal/pkg/utils	(cached)
?   	github.com/greenpau/caddy-auth-portal/pkg/validators	[no test files]
```

Let me know if you need more context, or if the implementation doesn't work for you. I'll be happy to discuss and revisit!

**Comment by greenpau** [comment]
@golgeek , thank you for the PR! I will educate myself on the subject matter, i.e. your explanation.

**Comment by greenpau** [comment]
@golgeek , looks straightforward! Do you need to do the same in JWT plugin!

**Comment by golgeek** [comment]
@greenpau I actually don't need it in the JWT plugin as I get all the info I need from the portal plugin.

But I'd be happy to add it in there too for harmonization, especially as the logic is exactly the same and it should take no more than 10 minutes!

**Comment by greenpau** [comment]
@golgeek , thank you for your contribution! Merged. Will do a release shortly.

**Comment by greenpau** [comment]
@golgeek , released [v1.4.15](https://github.com/greenpau/caddy-auth-portal/releases/tag/v1.4.15)!

> But I'd be happy to add it in there too for harmonization, especially as the logic is exactly the same and it should take no more than 10 minutes!

Yes, please! 😃 👍

**Comment by golgeek** [comment]
Thanks for merging and releasing so fast!
